code/game/objects/items/devices/translator.dm

```diff
@@ -46,34 +46,26 @@
 		return
 
 	var/mob/living/L = loc
+	if(visual && ((L.sdisabilities & BLIND) || L.eye_blind))
+		return
+	if(audio && ((L.sdisabilities & DEAF) || L.ear_deaf))
+		return
 
-	if(!language)
-		return //Borgs were causing runtimes when passing language=null
-
-	if (language && (language.flags & NONVERBAL))
-		return //Not gonna translate sign language
-
-	if (!language.machine_understands)
-		return //Any other languages that it can't translate.
-
-	if (visual && ((L.sdisabilities & BLIND) || L.eye_blind))
-		return //Can't see the screen, don't get the message
-
-	if (audio && ((L.sdisabilities & DEAF) || L.ear_deaf))
-		return //Can't hear the translation, don't get the message
-
-	//Only translate if they can't understand, otherwise pointlessly spammy
-	//I'll just assume they don't look at the screen in that case
-
-	//They don't understand the spoken language we're translating FROM
-	if(!L.say_understands(speaker,language))
-		//They understand the output language
-		if(L.say_understands(null,langset))
-			to_chat(L, "<i><b>[src]</b> translates, </i>\"<span class='[langset.colour]'>[message]</span>\"")
-
-		//They don't understand the output language
-		else
-			to_chat(L, "<i><b>[src]</b> translates, </i>\"<span class='[langset.colour]'>[langset.scramble(message)]</span>\"")
+	for(var/datum/multilingual_say_piece/S in message_pieces)
+		if(!S.speaking)
+			continue
+		if(S.speaking.flags & NONVERBAL)
+			continue
+		if(!S.speaking.machine_understands)
+			continue
+		// If our user doesn't understand the language being spoken, translate
+		if(!L.say_understands(M, S.speaking))
+			//They understand the output language
+			if(L.say_understands(null, langset))
```

Comment:
For a given message:
```
&h Hi, this is a demonstration &0 of the new multilingual say system. &4 With this system, you can now speak &0 any number of different languages &0 in one message &1 intersparsed throughout. &0 This message contains Akhani, Terminus, Sol Common, and Galactic Common.
```
Where the user understands &h and &0, but not &4 and &1, the person's chat log would look roughly like this:
```
Shadow Larkens chatters, "Hi, this is a demonstration of the new multilingual say system. <span>Garble grwarble frawble slar</span> any number of different languages in one message <span>zoop zabbity blah blu.</span> This message contains Akhnai, Terminus, Sol Common, and Galactic Common."
The translator translates, "With this system, you can now speak"
The translator translates, "intersparsed throughout."
```

It's not especially graceful, admittedly. I could probably have this reconstruct the whole message into one language and use hear_say, if that'd be preferable, but it'd be a bit of a change in behavior due to following hear_say rules instead of directly to_chat'ing the user.

Reply:
![https://i.tigercat2000.net/2020/04/dreamseeker_MkqorPO8YP.png](https://i.tigercat2000.net/2020/04/dreamseeker_MkqorPO8YP.png)

Ought to work.